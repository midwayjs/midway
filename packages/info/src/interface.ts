export type InfoValueType = 'html' | 'json';

export const DefaultHiddenKey = ['keys', '*key', '*token', '*secret*', 'pass*'];

export interface TypeInfo {
  type: string;
  info: {
    [key: string]: string | number;
  }
}

export interface InfoConfigOptions {
  title: string;
  infoPath: string;
  hiddenKey: Array<string>;
}
