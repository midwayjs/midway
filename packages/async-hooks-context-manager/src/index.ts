export * from './asyncHooksContextManager';
export * from './asyncLocalStorageContextManager';
export { isSemverGreaterThanOrEqualTo, createContextManager } from './util';
