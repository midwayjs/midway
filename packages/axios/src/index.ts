export { AutoConfiguration as Configuration } from './configuration';
export * from './interface';
export * from './serviceManager';
