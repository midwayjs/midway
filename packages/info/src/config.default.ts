import { DefaultHiddenKey, InfoConfigOptions } from './interface';

export const info = {
  title: 'Midway Info',
  infoPath: '/_info',
  hiddenKey: DefaultHiddenKey,
} as InfoConfigOptions;
