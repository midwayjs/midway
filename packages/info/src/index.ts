export { InfoConfiguration as Configuration } from './configuration';
export * from './infoService';
export * from './middleware/info.middleware';
export * from './interface';
