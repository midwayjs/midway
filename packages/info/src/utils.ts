import { TypeInfo } from './interface';

export function safeRequire(mod: string, defaultValue?) {
  try {
    return require(mod);
  } catch (e) {
    return defaultValue;
  }
}

export function bitToMB(bit: number): string {
  return Number((bit || 0) / 1024 / 1024).toFixed(2) + ' MB';
}

export function renderToHtml(infoList: TypeInfo[], title): string {
  let html = `<div style="padding: 24px; font-size: 36px;background: #9999cb;font-weight: bold;">${title}</div>`;
  html += infoList
    .map(info => {
      const infoList = Object.keys(info.info || {});
      return `<div class="infoType">${info.type}</div>
    <table style="border-collapse: collapse;width: 100%;">${infoList
      .map(infoName => {
        return `<tr style="border: 1px solid #666;"><td class="infoName">${infoName}</td><td class="infoValue">${info.info[infoName]}</td></tr>`;
      })
      .join('')}</table>`;
    })
    .join('');
  return `<style>
  .infoType {background: #8e8dc5;margin-top: 24px;padding: 12px;font-weight: bold;font-size: 16px;}
  .infoName {font-weight:bold;vertical-align: top;background:#c2c6fc;width: 200px;word-break:break-all;padding: 12px;}
  .infoValue {vertical-align: top;background:#d8d8d8;padding: 12px;word-break:break-all;}
  </style><div style="margin: 24px auto;max-width: 720px;min-width: 440px;">${html}</div>`;
}

export function safeContent(value = '') {
  if (value.length < 3) {
    return '*'.repeat(value.length);
  } else if (value.length < 6) {
    return value[0] + '*'.repeat(value.length - 1);
  } else if (value.length < 10) {
    return value[0] + '*'.repeat(value.length - 2) + value[value.length - 1];
  } else if (value.length < 15) {
    return value.slice(0, 2) + '*'.repeat(value.length - 4) + value.slice(-2);
  } else {
    return value.slice(0, 3) + '*'.repeat(value.length - 6) + value.slice(-3);
  }
}
