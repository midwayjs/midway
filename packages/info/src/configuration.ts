import {
  Configuration,
  Inject,
  MidwayApplicationManager,
} from '@midwayjs/core';
import { InfoMiddleware } from './middleware/info.middleware';
import * as DefaultConfig from './config.default';

@Configuration({
  namespace: 'info',
  importConfigs: [
    {
      default: DefaultConfig,
    },
  ],
})
export class InfoConfiguration {
  @Inject()
  applicationManager: MidwayApplicationManager;

  async onReady() {
    this.applicationManager
      .getApplications(['koa', 'faas', 'express', 'egg'])
      .forEach(app => {
        app.useMiddleware(InfoMiddleware);
      });
  }
}
