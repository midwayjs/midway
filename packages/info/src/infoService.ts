import {
  Provide,
  Scope,
  ScopeEnum,
  Inject,
  ApplicationContext,
  Config,
  Init,
} from '@midwayjs/core';
import {
  MidwayInformationService,
  IMidwayContainer,
  MidwayConfigService,
  MidwayEnvironmentService,
} from '@midwayjs/core';
import { InfoValueType, TypeInfo } from './interface';
import { bitToMB, renderToHtml, safeContent, safeRequire } from './utils';
import {
  hostname,
  homedir,
  cpus,
  networkInterfaces,
  uptime,
  totalmem,
} from 'os';
import { join } from 'path';
import * as pm from 'picomatch';

@Provide()
@Scope(ScopeEnum.Singleton)
export class InfoService {
  @Inject()
  midwayInformationService: MidwayInformationService;

  @Inject()
  configService: MidwayConfigService;

  @Inject()
  environment: MidwayEnvironmentService;

  @Config('info.title')
  titleConfig;

  @Config('info.hiddenKey')
  defaultHiddenKey: string[];

  secretMatchList: Array<any>;

  @ApplicationContext()
  container: IMidwayContainer;

  @Init()
  async init() {
    this.secretMatchList = Array.from(new Set(this.defaultHiddenKey)).map(
      pattern => {
        return pm(pattern);
      }
    );
  }

  info(infoValueType?: InfoValueType) {
    const info: TypeInfo[] = [];
    info.push(this.projectInfo());
    info.push(this.systemInfo());
    info.push(this.resourceOccupationInfo());
    info.push(this.softwareInfo());
    info.push(this.midwayConfig());
    info.push(this.midwayService());
    info.push(this.timeInfo());
    info.push(this.envInfo());
    info.push(this.dependenciesInfo());
    info.push(this.networkInfo());

    if (infoValueType === 'html') {
      return renderToHtml(info, this.titleConfig);
    }
    return info;
  }

  projectInfo(): TypeInfo {
    return {
      type: 'Project',
      info: {
        Project: this.midwayInformationService.getProjectName(),
        AppDir: this.midwayInformationService.getAppDir(),
        BaseDir: this.midwayInformationService.getBaseDir(),
        Root: this.midwayInformationService.getRoot(),
        Env: this.environment.getCurrentEnvironment(),
      },
    };
  }

  systemInfo(): TypeInfo {
    const _platform = process.platform;
    return {
      type: 'System',
      info: {
        Platform: _platform === 'win32' ? 'Windows' : _platform,
        Node: process.versions.node,
        V8: process.versions.v8,
        ProcessId: process.pid,
        Arch: process.arch,
        Hostname: hostname(),
        HomeDir: homedir(),
        CWD: process.cwd(),
        ExecCommand: [].concat(process.argv, process.execArgv).join(' '),
      },
    };
  }

  resourceOccupationInfo(): TypeInfo {
    const memory = process.memoryUsage();
    const cpu = cpus();
    return {
      type: 'Memory & CPU',
      info: {
        'Memory Total Occupy': bitToMB(memory.rss),
        'Heap Total Occupy': bitToMB(memory.heapTotal),
        'Heap Used': bitToMB(memory.heapUsed),
        'V8 C++ Object Memory': bitToMB(memory.external),
        'System Total Memory': bitToMB(totalmem()),
        CPU: `${cpu[0] ? `${cpu[0].model} ${cpu[0].speed}MHz` : ''} ${
          cpu.length
        } core `,
        'CPU Usage': cpu
          .map(cpuInfo => {
            const times = cpuInfo.times;
            return (
              (
                (1 -
                  times.idle /
                    (times.idle +
                      times.user +
                      times.nice +
                      times.sys +
                      times.irq)) *
                100
              ).toFixed(2) + '%'
            );
          })
          .join(' / '),
      },
    };
  }

  softwareInfo(): TypeInfo {
    const npmModuleList = [
      '@midwayjs/core',
      '@midwayjs/decorator',
      '@midwayjs/faas',
    ];
    const info = {};
    for (const modName of npmModuleList) {
      const modulePkg = safeRequire(join(modName, 'package.json'));
      if (modulePkg) {
        info[modName] = modulePkg.version;
      }
    }
    return {
      type: 'Software',
      info,
    };
  }

  envInfo(): TypeInfo {
    const env = {};
    Object.keys(process.env).forEach(envName => {
      env[envName] = this.filterSecretContent(envName, process.env[envName]);
    });
    return {
      type: 'Environment Variable',
      info: env,
    };
  }

  timeInfo(): TypeInfo {
    const t = new Date().toString().split(' ');
    return {
      type: 'Time',
      info: {
        Current: Date.now(),
        Uptime: uptime(),
        Timezone: t.length >= 7 ? t[5] : '',
        TimezoneName:
          t.length >= 7
            ? t.slice(6).join(' ').replace(/\(/g, '').replace(/\)/g, '')
            : '',
      },
    };
  }

  networkInfo(): TypeInfo {
    const net = networkInterfaces();
    const info = {};
    Object.keys(net).forEach(type => {
      const netItemList = net[type];
      let newType = type;
      if (type[type.length - 1] === '0') {
        newType = type.slice(0, -1);
      }
      // ignore localhost
      if (newType === 'lo') {
        return;
      }
      info[newType] = netItemList
        .sort(item => {
          if (item.family === 'IPv4') {
            return -1;
          }
          return 1;
        })
        .map(netItem => {
          return `${netItem.family} ${netItem.address}`;
        })
        .join(' / ');
    });
    return {
      type: 'Network',
      info,
    };
  }

  dependenciesInfo(): TypeInfo {
    const pkg = this.midwayInformationService.getPkg();
    const dependencies = pkg.dependencies || {};
    const info = {};
    Object.keys(dependencies).forEach(modName => {
      const modInfo = safeRequire(join(modName, 'package.json'), {});
      info[modName] = `${modInfo.version || 'Not Found'}(${
        dependencies[modName]
      })`;
    });
    return {
      type: 'Dependencies',
      info,
    };
  }

  midwayService() {
    const info = {};
    if (this.container?.registry) {
      for (const item of (this.container as any).registry) {
        const [key, value] = item;
        const name = value ? value?.name || value : typeof value;
        info[key] = `${value?.namespace ? `${value?.namespace}:` : ''}${name}${
          value?.scope ? ` [${value?.scope}]` : ''
        }`;
      }
    }

    return {
      type: 'Midway Service',
      info,
    };
  }

  midwayConfig() {
    const info = {};
    const config = this.configService.getConfiguration() || {};
    Object.keys(config).forEach(key => {
      info[key] = this.safeJson(this.filterSecretContent(key, config[key]));
    });
    return {
      type: 'Midway Config',
      info,
    };
  }

  protected filterSecretContent(key, value) {
    if (typeof value === 'string') {
      const find = this.secretMatchList.some(isMatch => {
        return isMatch(key.toLowerCase());
      });
      if (find) {
        return safeContent(value);
      }
    } else if (Array.isArray(value)) {
      return value.map(item => {
        return this.filterSecretContent(key, item);
      });
    }

    return value;
  }

  protected safeJson(value: any): string {
    switch (typeof value) {
      case 'string':
        return `"${value}"`;
      case 'number':
        return `${value}`;
      case 'boolean':
        return String(value);
      case 'object':
        if (!value) {
          return 'null';
        }
        if (Array.isArray(value)) {
          return `[${value.map(item => this.safeJson(item)).join(',')}]`;
        }
        if (value instanceof RegExp) {
          return `"${value.toString()}"`;
        }
        // eslint-disable-next-line no-case-declarations
        const props = [];
        for (const key in value) {
          props.push(
            `"${key}":${this.safeJson(
              this.filterSecretContent(key, value[key])
            )}`
          );
        }
        return `{${props.join(',')}}`;
      case 'function':
        return `function ${value.name}(${value.length} args)`;
    }
    return '';
  }
}
