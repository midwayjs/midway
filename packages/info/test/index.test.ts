import { close, createApp, createHttpRequest, createLightApp } from '@midwayjs/mock';
import { InfoService } from '../src';
import { join } from 'path';
import * as info from '../src';

describe('test/index.test.ts', () => {
  it('info service', async () => {
    const app = await createLightApp(join(__dirname, './fixtures/app'));
    const infoService = await app.getApplicationContext().getAsync(InfoService);
    expect(infoService).toBeDefined();
    const info = infoService.info();
    expect(Object.keys(info).length > 5).toBeTruthy();

    const html = infoService.info('html');
    expect(html).toMatch('Midway Info');
    await close(app);
  });

  it('test koa middleware', async () => {
    const app = await createApp(join(__dirname, './fixtures/web-koa'));
    const result = await createHttpRequest(app).get('/_test_route');
    expect(result.type).toEqual('text/html');
    expect(result.text).toMatch('xxx');
    await close(app);
  });

  it('test express middleware', async () => {
    const app = await createApp(join(__dirname, './fixtures/web-express'));
    const result = await createHttpRequest(app).get('/_info');
    expect(result.type).toEqual('text/html');
    expect(result.text).toMatch('Midway Info');
    await close(app);
  });

  it('should test get config and filter secret', async () => {
    const app = await createLightApp('', {
      globalConfig: {
        keys: ['123445555'],
        oss: {
          clients: {
            default: {
              accessKeyId: '123',
              accessKeySecret: 'fjdlsaf',
            }
          }
        },
        cos: {
          default: {
            SecretId: 'dfdsafdsaf',
            SecretKey: 'tewqtewqvc',
            SecurityToken: 'fjkdlsajfdsaf'
          }
        }
      },
      imports: [
        info
      ]
    });
    const infoService = await app.getApplicationContext().getAsync(InfoService);
    expect(infoService.midwayConfig()).toMatchSnapshot();
  });
});
