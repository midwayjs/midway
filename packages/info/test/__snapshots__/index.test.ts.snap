// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/index.test.ts should test get config and filter secret 1`] = `
{
  "info": {
    "asyncContextManager": "{"enable":false}",
    "cos": "{"default":{"SecretId":"df******af","SecretKey":"te******vc","SecurityToken":"fj*********af"}}",
    "debug": "{"recordConfigMergeOrder":true}",
    "info": "{"title":"Midway Info","infoPath":"/_info","hiddenKey":["k***","****","*****n","********","p****"]}",
    "keys": "["1*******5"]",
    "midwayLogger": "{"default":{"dir":"logs","level":"info","consoleLevel":"info","auditFileDir":".audit","disableFile":true,"disableError":true},"clients":{"coreLogger":{"fileLogName":"midway-core.log"},"appLogger":{"fileLogName":"midway-app.log","aliasName":"logger"}}}",
    "oss": "{"clients":{"default":{"accessKeyId":"123","accessKeySecret":"f*****f"}}}",
  },
  "type": "Midway Config",
}
`;
